include/sound/soc-topology.h

```diff
@@ -44,6 +44,7 @@ enum snd_soc_dobj_type {
 	SND_SOC_DOBJ_DAI_LINK,
 	SND_SOC_DOBJ_CODEC_LINK,
 	SND_SOC_DOBJ_WIDGET,
+	SND_SOC_DOBJ_BE_DAI_LINK,
```

Comment:
Hum. If I look at include/uapi/sound/asoc.h, I see the following definitions

````c
#define SND_SOC_TPLG_TYPE_MIXER		1
#define SND_SOC_TPLG_TYPE_BYTES		2
#define SND_SOC_TPLG_TYPE_ENUM		3
#define SND_SOC_TPLG_TYPE_DAPM_GRAPH	4
#define SND_SOC_TPLG_TYPE_DAPM_WIDGET	5
#define SND_SOC_TPLG_TYPE_DAI_LINK	6
#define SND_SOC_TPLG_TYPE_PCM		7
#define SND_SOC_TPLG_TYPE_MANIFEST	8
#define SND_SOC_TPLG_TYPE_CODEC_LINK	9
#define SND_SOC_TPLG_TYPE_BACKEND_LINK	10
#define SND_SOC_TPLG_TYPE_PDATA		11
#define SND_SOC_TPLG_TYPE_DAI		12
````

It looks like there was a correlation between the two structures initially, so should we use BACKEND_LINK instead? @mengdonglin can you comment on this?

Reply:
@plbossart @mengdonglin I have a question here.  We create a dobj with SND_SOC_DOBJ_DAI_LINK type for tplg type SND_SOC_TPLG_TYPE_PCM. 
[case SND_SOC_TPLG_TYPE_PCM](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/soc-topology.c#L2410)
```
soc_tplg_pcm_elems_load() -> soc_tplg_pcm_create() -> soc_tplg_fe_link_create()
```
[link->dobj.type = SND_SOC_DOBJ_DAI_LINK](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/soc-topology.c#L1792)

And config physical link on tplg type SND_SOC_TPLG_TYPE_DAI_LINK and SND_SOC_TPLG_TYPE_BACKEND_LINK.
[case SND_SOC_TPLG_TYPE_DAI_LINK:
case SND_SOC_TPLG_TYPE_BACKEND_LINK:](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/soc-topology.c#L2414)
It seems the tplg type and dobj type is mismatched already.
Should we fix it first?

Reply:
@plbossart If we find a SND_SOC_TPLG_TYPE_PCM type tplg when we parse a topology file, we will create a SND_SOC_DOBJ_PCM type dobj AND a SND_SOC_DOBJ_DAI_LINK type dobj. Please see the chart below. 
````
	soc_tplg_load_header()
	case SND_SOC_TPLG_TYPE_PCM:
                -> soc_tplg_pcm_elems_load ()
			-> soc_tplg_pcm_create()
				-> soc_tplg_dai_create()
						dai_drv->dobj.type = SND_SOC_DOBJ_PCM;
						list_add(&dai_drv->dobj.list, &tplg->comp->dobj_list);
				-> soc_tplg_fe_link_create()
						link->dobj.type = SND_SOC_DOBJ_DAI_LINK;
						list_add(&link->dobj.list, &tplg->comp->dobj_list);
````
So it is not one to one mapping. 
Also, SND_SOC_TPLG_TYPE_DAI_LINK and SND_SOC_TPLG_TYPE_BACKEND_LINK are both using soc_tplg_link_elems_load() in soc_tplg_load_header(). So there is no different in dealing the two tplg type. My suggestion is as below. But it is a big change, we have to check and modify the topology file as well.
````
@@ -2427,10 +2427,10 @@ static int soc_tplg_load_header(struct soc_tplg *tplg,
        case SND_SOC_TPLG_TYPE_DAPM_WIDGET:
                return soc_tplg_dapm_widget_elems_load(tplg, hdr);
        case SND_SOC_TPLG_TYPE_PCM:
+       case SND_SOC_TPLG_TYPE_DAI_LINK:
                return soc_tplg_pcm_elems_load(tplg, hdr);
        case SND_SOC_TPLG_TYPE_DAI:
                return soc_tplg_dai_elems_load(tplg, hdr);
-       case SND_SOC_TPLG_TYPE_DAI_LINK:
        case SND_SOC_TPLG_TYPE_BACKEND_LINK:
                /* physical link configurations */
                return soc_tplg_link_elems_load(tplg, hdr);
````